src/librustc_mir/borrow_check/mod.rs

```diff
@@ -1837,17 +1839,43 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
             Place::Projection(box Projection {
                 base: Place::Local(local),
                 elem: ProjectionElem::Deref,
-            }) if self.mir.local_decls[*local].is_nonref_binding() =>
-            {
-                let (err_help_span, suggested_code) =
-                    find_place_to_suggest_ampmut(self.tcx, self.mir, *local);
-                err.span_suggestion(
-                    err_help_span,
-                    "consider changing this to be a mutable reference",
-                    suggested_code,
-                );
-
+            }) if self.mir.local_decls[*local].is_user_variable.is_some() => {
                 let local_decl = &self.mir.local_decls[*local];
+                let suggestion = match local_decl.is_user_variable {
+                    Some(ClearCrossCrate::Set(mir::BindingForm::ImplicitSelf)) => {
+                        Some(suggest_ampmut_self(local_decl))
+                    },
+
+                    Some(ClearCrossCrate::Set(mir::BindingForm::Var(mir::VarBindingForm {
+                        binding_mode: ty::BindingMode::BindByValue(_),
+                        opt_ty_info,
+                        ..
+                    }))) => Some(suggest_ampmut(
+                        self.tcx,
+                        self.mir,
+                        *local,
+                        local_decl,
+                        opt_ty_info,
+                    )),
+
+                    Some(ClearCrossCrate::Set(mir::BindingForm::Var(mir::VarBindingForm {
+                        binding_mode: ty::BindingMode::BindByReference(_),
+                        ..
+                    }))) => suggest_ref_mut(self.tcx, local_decl),
+
+                    Some(ClearCrossCrate::Clear) => bug!("saw cleared local state"),
+
+                    None => bug!(),
```

Comment:
Not really sure what the most idiomatic way to do this is. I could add `.as_ref().unwrap()` on line 1844, but that felt kinda ugly.